### PR TITLE
Anti did field map

### PR DESCRIPTION
**Pull request by luisaleperez**
BEGINRELEASENOTES
- implemented FieldMapXYZ.cpp 
     - a 3D field map defined on a grid in (x,y,z) 
- added field maps for the ILD solenoid and anti-DID
     -  fieldmaps/ild_fieldMap_antiDID_10cm_v1_20170223.root 
     -  fieldmaps/ild_fieldMap_Solenoid3.5T_StandardYoke_10cm_v1_20170223.root 
- added example configuration for using these maps to ILD_o1_v05

ENDRELEASENOTES

### CODE REVIEW

**Comment by petricm** [comment]
Dear @luisaleperez please write release notes.

**Comment by gaede** [comment]
@luisaleperez you can simply add the release notes by editing  the comment field for this PR between the BEGINRELEASENOTES/ENDRELEASNOTES tags

**Comment by luisaleperez** [comment]
BEGINRELEASENOTES

Merged Magnetic field bug fix (see issue https://github.com/iLCSoft/lcgeo/issues/65), with lcgeo branch AntiDIDFieldMap. This lcgeo branch was intended for the implementation of the anti-DID field map reader.

The latest calculation of both the solenoid and anti-DID field maps have been included inside the fieldmaps directory and have been used in the xml detector model file ILD/compact/ILD_o1_v05/ILD_o1_v05.xml

ENDRELEASENOTES

**Comment by petricm** [comment]
Just for future reference. The release notes should be in the initial comment I have moved them up. You can always edit your own comments.

**Comment by gaede** [comment]
@luisaleperez I have updated your release notes to make it clearer what you did. There is no point in referring to a feature branch as that never will make it into the main repository.
Also, could you please change ILD_o1_v05, so that it actually uses the old solenoid field and has the field maps only commented out as examples ?
We need to think how to best create models that use the field maps, but the default should be 'constant' B-field, as this is what is used by the standard reconstruction

**Comment by gaede** [comment]
@luisaleperez there seem to be conflicts in your PR. Did you run a rebase with the master before pushing to your branch ? Please resolve the conflicts **without** using the web interface, as this has caused problems in the past.
@andresailer can you provide the necessary commands to @luisaleperez for resolvin this w/o the web interface ?

**Comment by luisaleperez** [comment]
Done

**Comment by luisaleperez** [comment]
Frank, I followed your instructions running a rebase with the master before pushing the branch

**Comment by gaede** [comment]
@luisaleperez hmm, then something must have gone wrong somewhere (maybe when resolving conflicts ? ). 

In any case you seem to actually have missed the bug fix for overlaying the fields:
https://github.com/luisaleperez/lcgeo/blob/AntiDIDFieldMap/detector/other/FieldMapXYZ.cpp#L105
needs to be really:
```
globalField[0] += B;
```
and the same for the other coordinates...

Lets wait for @andresailer to provide the best commands to resolve the conflict...

**Comment by luisaleperez** [comment]
Doesn't the line 

globalField[0] += B;

has to be added for all the field maps? 

If we do in this way it means that the field will depend on the other the field maps are specified. Is there a way to initialize the field to zero and then add contributions?

**Comment by gaede** [comment]
It is done like that in all the field maps:
https://github.com/AIDASoft/DD4hep/pull/140/files
and 
https://github.com/iLCSoft/lcgeo/pull/67/commits/3c17beb215e28dc700b65cd3429976edd378f4c5

**Comment by andresailer** [comment]
See for pulling the new master and rebasing:
https://github.com/andresailer/tutorial/blob/master/README.md#incorporating-upstream-changes-after-creating-the-branch-rebasing

I have just added an explanation for conflict resolution using this branch/conflict as the exaple:
https://github.com/andresailer/tutorial/blob/master/README.md#mergerebase-conflict

**Comment by luisaleperez** [comment]
Hello I followed the instruction and solved the conflicts, but when I do 

git push downstream AntiDIDFieldMap

I get this. 

To https://luisaleperez@github.com/luisaleperez/lcgeo.git
 ! [rejected]        AntiDIDFieldMap -> AntiDIDFieldMap (non-fast-forward)
error: failed to push some refs to 'https://luisaleperez@github.com/luisaleperez/lcgeo.git'
To prevent you from losing history, non-fast-forward updates were rejected
Merge the remote changes before pushing again.  See the 'Note about
fast-forwards' section of 'git push --help' for details.

Do you have any suggestions?

**Comment by andresailer** [comment]
Please read:
https://github.com/andresailer/tutorial/blob/master/README.md#push-you-local-changes-to-your-downstream-remote

**Comment by luisaleperez** [comment]
Sorry, I missed the -f option in the push command. It is ok now.

**Comment by luisaleperez** [comment]
One thing. I tried to compile the new lcgeo code by doing this,

source /afs/desy.de/project/ilcsoft/sw/x86_64_gcc49_sl6/HEAD-2017-03-24/init_ilcsoft.sh

mkdir build; cd build

export CC=gcc
export CXX=g++
cmake -C /afs/desy.de/project/ilcsoft/sw/x86_64_gcc49_sl6/v01-19-01/ILCSoft.cmake ..

make clean
make -j2 install

and I get the following error. 

/afs/desy.de/user/l/laperez/ILC_Developments/6Mar2017/ILCSoft/lcgeo/HEAD-2017–02-23/detector/tracker/SiTrackerEndcap_o2_v02ext_geo.cpp:27:32: fatal error: UTIL/LCTrackerConf.h: No such file or directory
 #include "UTIL/LCTrackerConf.h"
                                ^
compilation terminated.
make[2]: *** [CMakeFiles/lcgeo.dir/detector/tracker/SiTrackerEndcap_o2_v02ext_geo.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/lcgeo.dir/all] Error 2
make: *** [all] Error 2

File: detector/other/FieldMapXYZ.cpp [hunk 1]
```diff
@@ -38,94 +38,125 @@ FieldMapXYZ::FieldMapXYZ() {
...
   //this area is covered, or add some more parameters for the values where the
   //field is supposed to be. Now we just assume it starts at 0/0/0, outside there is no field
-  if (not ( 0.0 <= x && x <= rhoMax &&
-	    0.0 <= y && y <= zMax ) ) {
-
+  if (not ( x >= xMin && x <= xMax &&
+            y >= yMin && y <= yMax &&
+            z >= zMin && z <= zMax )
+     ) {
+    globalField[0] = globalField[1] = globalField[2] = 0.0;
```
Comment (andresailer) [inline detector/other/FieldMapXYZ.cpp]: This line should be removed again.

**Comment by gaede** [comment]
@luisaleperez can you please remove the line that sets the field to zero. You need to create a new PR as this one is already closed.

**Pull request merged**